d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px; height: 163px">
</div>

# CSE Coding Assignment
## Instructions

- Please answer all questions
- You can use any language you wish (e.g. Python, Scala, SQL...)
- Several Markdown cells require completion. Please edit the Markdown cells to include your answer.
- Your final notebook should compile without errors when you click "Run All"

**Please do not publish questions. This is a confidential assignment.**

### Creating a Cluster

You will need to create a Databricks Cluster. More information on this process is available here: https://docs.databricks.com/user-guide/clusters/create.html

## Getting Started

**REQUIRED:** Run the following cells exactly as written to retrieve the necessary Coding Assignment Data Sets from Amazon S3.

In [0]:
%sh curl --remote-name-all 'https://files.training.databricks.com/assessments/cse-take-home/{covertype,kafka,treecover,u.data,u.item}.csv'


[1/5]: https://files.training.databricks.com/assessments/cse-take-home/covertype.csv --> covertype.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   133  100   133    0     0    773      0 --:--:-- --:--:-- --:--:--   773

[2/5]: https://files.training.databricks.com/assessments/cse-take-home/kafka.csv --> kafka.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  135k  100  135k    0     0   863k      0 --:--:-- --:--:-- --:--:--  863k

[3/5]: https://files.training.databricks.com/assessments/cse-take-home/treecover.csv --> treecover.csv
  % Total    % Received % Xferd  Average Speed   Time    Time 

In [0]:
dbutils.fs.cp("file:/databricks/driver/covertype.csv", "dbfs:/FileStore/tmp/covertype.csv")
dbutils.fs.cp("file:/databricks/driver/kafka.csv", "dbfs:/FileStore/tmp/kafka.csv")
dbutils.fs.cp("file:/databricks/driver/treecover.csv", "dbfs:/FileStore/tmp/treecover.csv")
dbutils.fs.cp("file:/databricks/driver/u.data.csv", "dbfs:/FileStore/tmp/u.data.csv")
dbutils.fs.cp("file:/databricks/driver/u.item.csv", "dbfs:/FileStore/tmp/u.item.csv")

Out[2]: True

## Part 1: Reading and Parsing Data

### Question 1:  Code Challenge - Load a CSV

- Load the CSV file at `dbfs:/FileStore/tmp/nl/treecover.csv` into a DataFrame.
- Use Apache Spark to read in the data, assigned to the variable `treeCoverDF`.
- Your method to get the CSV file into Databricks isn't graded. We are only concerned with how you use Spark to parse and load the actual data. 
- Please use the `inferSchema` option.

In [0]:
# YOUR CODE HERE
treeCoverDF=spark.read.csv("/FileStore/tmp/treecover.csv",inferSchema="true",header="true")
treeCoverDF.show()

+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
| Id|Elevation|Aspect|Slope|Horizontal_Distance_To_Hydrology|Vertical_Distance_To_Hydrology|Horizontal_Distance_To_Roadways|Horizontal_Distance_To_Fire_Points|Cover_Type|Soil_Type|Wilderness_Area|Hillshade|
+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
|  1|     2596|    51|    3|                             258|                             0|                            510|                              6279|         5|       29|              1|      9am|
|  2|     2590|    56|    2|                             212|                            -6|                            390|                              6225|         5|  

### Question 2:  Code Challenge - Print the Schema

Use Apache Spark to display the Schema of the `treeCoverDF` Dataframe.

In [0]:
treeCoverDF.printSchema()

root
 |-- Id: integer (nullable = true)
 |-- Elevation: integer (nullable = true)
 |-- Aspect: integer (nullable = true)
 |-- Slope: integer (nullable = true)
 |-- Horizontal_Distance_To_Hydrology: integer (nullable = true)
 |-- Vertical_Distance_To_Hydrology: integer (nullable = true)
 |-- Horizontal_Distance_To_Roadways: integer (nullable = true)
 |-- Horizontal_Distance_To_Fire_Points: integer (nullable = true)
 |-- Cover_Type: integer (nullable = true)
 |-- Soil_Type: integer (nullable = true)
 |-- Wilderness_Area: integer (nullable = true)
 |-- Hillshade: string (nullable = true)



### Question 3:  Code Challenge - Rows & Columns

Use Apache Spark to display the number of rows and columns in the DataFrame.

In [0]:
print("Rows:",treeCoverDF.count())
print("Columns:",len(treeCoverDF.columns))

Rows: 15120
Columns: 12


#Part 2: Analysis

### Question 4:  Code Challenge - Summary Statistics for a Feature

Use Apache Spark to answer these questions about the `treeCoverDF` DataFrame:
- What is the range - minimum and maximum - of values for the feature `elevation`?
- What are the mean and standard deviation of the feature `elevation`?

In [0]:
from pyspark.sql.functions import *
treeCoverDF.agg(min(col('elevation')),max(col('elevation'))).show()
treeCoverDF.agg(mean(col('elevation')),stddev(col('elevation'))).show()

+--------------+--------------+
|min(elevation)|max(elevation)|
+--------------+--------------+
|          1863|          3849|
+--------------+--------------+

+------------------+----------------------+
|    avg(elevation)|stddev_samp(elevation)|
+------------------+----------------------+
|2749.3225529100528|    417.67818734804985|
+------------------+----------------------+



### Answer #4:

- Min `elevation`: `1863`
- Max `elevation`: `3849`
- Mean `elevation`: `2749`
- Standard Deviation of `elevation`: `417`

### Question 5:  Code Challenge - Record Count

Use Apache Spark to answer the following question:
- How many entries in the dataset have an `elevation` greater than or equal to 2749.32 meters **AND** a `Cover_Type` of 1 or 2?

In [0]:
treeCoverDF.filter('elevation >= 2749.32 and cover_type =1 or cover_type = 2').show()

+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
| Id|Elevation|Aspect|Slope|Horizontal_Distance_To_Hydrology|Vertical_Distance_To_Hydrology|Horizontal_Distance_To_Roadways|Horizontal_Distance_To_Fire_Points|Cover_Type|Soil_Type|Wilderness_Area|Hillshade|
+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
|  3|     2804|   139|    9|                             268|                            65|                           3180|                              6121|         2|       12|              1|      3pm|
|  4|     2785|   155|   18|                             242|                           118|                           3090|                              6211|         2|  

### Question 6: Code Challenge - Compute a Percentage

Use Apache Spark to answer the following question:
- What percentage of entries with `Cover_Type` 1 or 2 have an `elevation` at or above 2749.32 meters?

In [0]:
filter_count=treeCoverDF.filter('elevation >=2749.32 and cover_type = 1 or cover_type = 2').count()
total_rec=treeCoverDF.count()
Percentage=(filter_count*100)/total_rec
print(Percentage)

28.30026455026455


### Question 7: Code Challenge - Visualize Feature Distribution

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart that helps visualize the distribution of different Wilderness Areas in our dataset

In [0]:
display(treeCoverDF.select('wilderness_area').distinct())

wilderness_area 1 3 4 2

### Question 8: Code Challenge - Visualize Average Elevation by Cover Type 

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart showing the average elevation of each cover type with string labels for cover type

**NOTE: you will need to match the integer values in the column `treeCoverDF.Cover_Type` to the string values in `dbfs:/FileStore/tmp/nl/covertype.csv` to retrieve the Cover Type Labels. It is recommended to use an Apache Spark join.**

In [0]:
Cover_type=spark.read.csv("dbfs:/FileStore/tmp/covertype.csv",header="true", inferSchema="true")
Cover_type.show()
newdf=treeCoverDF.join(Cover_type,treeCoverDF.Cover_Type==Cover_type.cover_type_key)
display(newdf.select('elevation','cover_type_label'))

+--------------+-----------------+
|cover_type_key| cover_type_label|
+--------------+-----------------+
|             1|       Spruce/Fir|
|             2|   Lodgepole Pine|
|             3|   Ponderosa Pine|
|             4|Cottonwood/Willow|
|             5|            Aspen|
|             6|      Douglas-fir|
|             7|        Krummholz|
+--------------+-----------------+



---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<command-3725234820062639> in <module>
      1 Cover_type=spark.read.csv("dbfs:/FileStore/tmp/covertype.csv",header="true", inferSchema="true")
      2 Cover_type.show()
----> 3 newdf=treeCoverDF.join(Cover_type,treeCoverDF.Cover_Type==Cover_type.cover_type_key)
      4 display(newdf.select('elevation','cover_type_label'))

NameError: name 'treeCoverDF' is not defined

#Part 3: Data Ingestion, Cleansing, and Transformations

## Instructions 

This is a multi-step, data pipeline question in which you need to achieve a few objectives to build a successful job.

### Data Sets

#### `u.data.csv`

- The full u data set, 100000 ratings by 943 users on 1682 items. 
- Each user has rated at least 20 movies.  
- Users and items are numbered consecutively from 1. 
- The data is randomly ordered. 
- This is a tab separated file consisting of four columns: 
   - user id 
   - movie id 
   - rating 
   - date (unix seconds since 1/1/1970 UTC)

#### Desired schema

- `user_id INTEGER`
- `movie_id INTEGER`
- `rating INTEGER`
- `date DATE `

#### `u.item.csv`

- This is a `|` separated file consisting of six columns:
   - movie id
   - movie title
   - release date
   - video release date
   - IMDb URL
   - genre
- movie ids in this file match movie ids in `u.data`.

#### Desired schema

- `movie_id INTEGER`
- `movie_title STRING`

### Question 9:  Code Challenge - Load DataFrames

Use Apache Spark to perform the following:
1. define the correct schemas for each Data Set to be imported as described above  
   **note:** 
      - for `u.data.csv`, `date` *must* be stored using `DateType` with the format `yyyy-MM-dd`
      - you may need to ingest `timestamp` data using `IntegerType`
      - be sure to drop unneccesary columns for `u.item.csv`
1. import the two files as DataFrames names `uDataDF` and `uItemDF` using the schemas you defined and these paths:
   - `dbfs:/FileStore/tmp/u.data.csv`
   - `dbfs:/FileStore/tmp/u.item.csv`
1. order the `uDataDF` DataFrame by the `date` column

**NOTE:** Please display the DataFrames, `uDataDF` and `uItemDF` after loading.

#### `uDataDF`

In [0]:
from pyspark.sql.functions  import date_format

uDataDF=spark.read.load("dbfs:/FileStore/tmp/u.data.csv",format="csv",sep="\t",header="false",inferSchema="true")

uDataDF = uDataDF.withColumnRenamed("_c0", "user_id").withColumnRenamed("_c1", "movie_id").withColumnRenamed("_c2", "rating").withColumnRenamed("_c3", "date").withColumn("tsDate",from_unixtime("date")).withColumn("newdate",date_format('tsDate', "yyyy-MM-dd")).drop('tsDate','date')

uDataDF.sort(desc("newdate")).show()

#from_unixtime(col("timestamp_3"),"MM-dd-yyyy").alias("timestamp_3")


+-------+--------+------+----------+
|user_id|movie_id|rating|   newdate|
+-------+--------+------+----------+
|    189|      30|     4|1998-04-22|
|    189|     496|     5|1998-04-22|
|    189|     120|     1|1998-04-22|
|    189|     520|     5|1998-04-22|
|    189|     462|     5|1998-04-22|
|    189|     526|     4|1998-04-22|
|    159|     259|     4|1998-04-22|
|    189|      20|     5|1998-04-22|
|    189|     197|     5|1998-04-22|
|    189|     694|     4|1998-04-22|
|    189|      45|     3|1998-04-22|
|    189|     751|     4|1998-04-22|
|    189|     276|     3|1998-04-22|
|    189|    1115|     4|1998-04-22|
|    189|     253|     4|1998-04-22|
|    189|       1|     5|1998-04-22|
|    189|     480|     5|1998-04-22|
|    189|    1060|     5|1998-04-22|
|    189|     705|     4|1998-04-22|
|    189|      91|     3|1998-04-22|
+-------+--------+------+----------+
only showing top 20 rows



#### `uItemDF`

In [0]:
uItemDF=spark.read.load("dbfs:/FileStore/tmp/u.item.csv",format="csv",sep="|",header="false",inferSchema="true")
uItemDF=uItemDF.select(col("_c0").alias("movie_id"),col("_c1").alias("movie_title"),col("_c2").alias("release_date"),col("_c3").alias("video_release_date"),col("_c4").alias("imd_url"),col("_c5").alias("genre"))
uItemDF.show()

+--------+--------------------+------------+------------------+--------------------+-----+
|movie_id|         movie_title|release_date|video_release_date|             imd_url|genre|
+--------+--------------------+------------+------------------+--------------------+-----+
|       1|    Toy Story (1995)| 01-Jan-1995|              null|http://us.imdb.co...|    0|
|       2|    GoldenEye (1995)| 01-Jan-1995|              null|http://us.imdb.co...|    0|
|       3|   Four Rooms (1995)| 01-Jan-1995|              null|http://us.imdb.co...|    0|
|       4|   Get Shorty (1995)| 01-Jan-1995|              null|http://us.imdb.co...|    0|
|       5|      Copycat (1995)| 01-Jan-1995|              null|http://us.imdb.co...|    0|
|       6|Shanghai Triad (Y...| 01-Jan-1995|              null|http://us.imdb.co...|    0|
|       7|Twelve Monkeys (1...| 01-Jan-1995|              null|http://us.imdb.co...|    0|
|       8|         Babe (1995)| 01-Jan-1995|              null|http://us.imdb.co...|    0|

### Question 10:  Code Challenge - Perform a Join

Use Apache Spark to do the following:
- join `uDataDF` and `uItemDf` on `movie_id` as a new DataFrame called `uMovieDF`  
   **note:** make sure you do not create duplicate `movie_id` columns
   
**NOTE:** Please display the DataFrame `uMovieDF`.

In [0]:
uMovieDF=uDataDF.join(uItemDF,['movie_id'])
uMovieDF.show()

+--------+-------+------+----------+--------------------+------------+------------------+--------------------+-----+
|movie_id|user_id|rating|   newdate|         movie_title|release_date|video_release_date|             imd_url|genre|
+--------+-------+------+----------+--------------------+------------+------------------+--------------------+-----+
|     242|    196|     3|1997-12-04|        Kolya (1996)| 24-Jan-1997|              null|http://us.imdb.co...|    0|
|     302|    186|     3|1998-04-04|L.A. Confidential...| 01-Jan-1997|              null|http://us.imdb.co...|    0|
|     377|     22|     1|1997-11-07| Heavyweights (1994)| 01-Jan-1994|              null|http://us.imdb.co...|    0|
|      51|    244|     2|1997-11-27|Legends of the Fa...| 01-Jan-1994|              null|http://us.imdb.co...|    0|
|     346|    166|     1|1998-02-02| Jackie Brown (1997)| 01-Jan-1997|              null|http://us.imdb.co...|    0|
|     474|    298|     4|1998-01-07|Dr. Strangelove o...| 01-Jan

### Question 11:  Code Challenge - Perform an Aggregation

Use Apache Spark to do the following:
1. create an aggregate DataFrame, `aggDF` by
  1. extracting the year from the `date` (of the review)
  1. getting the average rating of each film per year as a column named `average_rating`
  1. ordering descending by year and average rating
1. write the resulting dataframe to a table named "movie_by_year_average_rating" in the Default database  
   **note:** use `mode(overwrite)` 

#### Desired Schema
The schema of you resulting DataFrame should be:
- `year INTEGER`
- `movie_title STRING`
- `average_rating DOUBLE`

**NOTE:** Please display the DataFrame `aggDF`.

In [0]:
from pyspark.sql.functions import year
from pyspark.sql.window import Window
uMovieDF1 = uMovieDF.withColumn("year",year("newdate"))


windowSpec = Window.partitionBy(uMovieDF1['year'])
windowSpec.show()

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<command-4475868478686193> in <module>
      5 
      6 windowSpec = Window.partitionBy(uMovieDF1['year'])
----> 7 windowSpec.show()

AttributeError: 'WindowSpec' object has no attribute 'show'

In [0]:
from pyspark.sql.functions import year
from pyspark.sql.window import Window
uMovieDF1 = uMovieDF.withColumn("year",year("newdate"))


windowSpec = Window.partitionBy(uMovieDF1['year'])
cols = ["year","movie_title"]


uMovieDF1.withColumn('average_rating', avg("rating").over(Window.partitionBy(cols))).drop("movie_id","user_id","rating","newdate","release_date","video_release_date","imdb_url","genre").distinct().show()

+--------------------+--------------------+----+------------------+
|         movie_title|             imd_url|year|    average_rating|
+--------------------+--------------------+----+------------------+
|'Til There Was Yo...|http://us.imdb.co...|1997|               1.5|
|        1-900 (1994)|http://us.imdb.co...|1997|               3.0|
|101 Dalmatians (1...|http://us.imdb.co...|1997|2.8358208955223883|
| 12 Angry Men (1957)|http://us.imdb.co...|1997| 4.295774647887324|
|          187 (1997)|http://us.imdb.co...|1997| 2.857142857142857|
|2 Days in the Val...|http://us.imdb.co...|1997|              3.25|
|20,000 Leagues Un...|http://us.imdb.co...|1997|3.4634146341463414|
|2001: A Space Ody...|http://us.imdb.co...|1997|               4.0|
|39 Steps, The (1935)|http://us.imdb.co...|1997| 3.896551724137931|
|        8 1/2 (1963)|http://us.imdb.co...|1997|              3.75|
|8 Heads in a Duff...|http://us.imdb.co...|1997|               1.0|
|    8 Seconds (1994)|http://us.imdb.co...|1997|

## Part 4: Fun with JSON

JSON values are typically passed by message brokers such as Kafka or Kinesis in a string encoding. When consumed by a Spark Structured Streaming application, this json must be converted into a nested object in order to be used.

Below is a list of json strings that represents how data might be passed from a message broker.

**Note:** Make sure to run the cell below to retrieve the sample data.

In [0]:
%python


sampleJson = [
 ('{"user":100, "ips" : ["191.168.192.101", "191.168.192.103", "191.168.192.96", "191.168.192.99"]}',), 
 ('{"user":101, "ips" : ["191.168.192.102", "191.168.192.105", "191.168.192.103", "191.168.192.107"]}',), 
 ('{"user":102, "ips" : ["191.168.192.105", "191.168.192.101", "191.168.192.105", "191.168.192.107"]}',), 
 ('{"user":103, "ips" : ["191.168.192.96", "191.168.192.100", "191.168.192.107", "191.168.192.101"]}',), 
 ('{"user":104, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.102", "191.168.192.99"]}',), 
 ('{"user":105, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.100", "191.168.192.96"]}',), 
]

### Question 12:  Code Challenge - Count the IPs

Use any coding techniques known to you to parse this list of JSON strings to answer the following question:
- how many occurrences of each IP address are in this list?

#### Desired Output
Your results should be this:


| ip | count |
|:-:|:-:|
| `191.168.192.96` | `3` |
| `191.168.192.99` | `6` |
| `191.168.192.100` | `2` |
| `191.168.192.101` | `3` |
| `191.168.192.102` | `2` |
| `191.168.192.103` | `2` |
| `191.168.192.105` | `3` |
| `191.168.192.107` | `3` |

**NOTE:** The order of your results is not important.

In [0]:
from pyspark.sql.functions import explode
from pyspark.sql.functions import *
from pyspark.sql.types import *
df1=spark.createDataFrame(sampleJson)

sch=StructType([StructField('user',StringType(),False),StructField('ips',ArrayType(StringType()))])
df2=df1.withColumn("n",from_json(col("_1"),sch)).select("n.*")
df1.withColumn("n",from_json(col("_1"),sch)).select("n.*").printSchema()
df2.withColumn('ip',explode("ips")).groupby('ip').agg(count("*").alias("count")).show()


root
 |-- user: string (nullable = true)
 |-- ips: array (nullable = true)
 |    |-- element: string (containsNull = true)

+---------------+-----+
|             ip|count|
+---------------+-----+
| 191.168.192.96|    3|
|191.168.192.101|    3|
|191.168.192.103|    2|
| 191.168.192.99|    6|
|191.168.192.105|    3|
|191.168.192.107|    3|
|191.168.192.102|    2|
|191.168.192.100|    2|
+---------------+-----+



+---------------+
|            col|
+---------------+
|191.168.192.101|
|191.168.192.103|
| 191.168.192.96|
| 191.168.192.99|
|191.168.192.102|
|191.168.192.105|
|191.168.192.103|
|191.168.192.107|
|191.168.192.105|
|191.168.192.101|
|191.168.192.105|
|191.168.192.107|
| 191.168.192.96|
|191.168.192.100|
|191.168.192.107|
|191.168.192.101|
| 191.168.192.99|
| 191.168.192.99|
|191.168.192.102|
| 191.168.192.99|
+---------------+
only showing top 20 rows



## Part 5: The Databricks API

### Question 13: Conceptual Question - the Databricks API

In 4-5 sentences, please explain what the Databricks API is used for at a high-level.

### Answer:

`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

### Question 14: Conceptual Question - Explain an API Call

In 4-5 sentences, please explain what this API call. Be sure to discuss some key attributes about the cluster.

### Answer:

The databricks API can be used for multiple purposes

Clusters API is used for managing the databricks spark clusters e.g. create, edit,. start, restart,delete and other cluster lifecyle calls`
DBFS API is used to interact with various data sources`
Groups API is used to manage groups of users.`
Likewise there are more API like Jobs, ML FLow,SCIM, Secrets, Libraries,Token ,Workspace to interact with various entities of Databricks`

## Part 6: Security

### Question 15: Conceptual Question - Security on Databricks

$ curl -n -X POST -H 'Content-Type: application/json'                      \
  -d '{                                                                     \
  "cluster_name": "high-concurrency-cluster",                               \
  "spark_version": "4.2.x-scala2.11",                                       \
  "node_type_id": "i3.xlarge",                                              \
  "spark_conf":{                                                            \
        "spark.databricks.cluster.profile":"serverless",                    \
        "spark.databricks.repl.allowedLanguages":"sql,python,r"             \
     },                                                                     \
     "aws_attributes":{                                                     \
        "zone_id":"us-west-2c",                                             \
        "first_on_demand":1,                                                \
        "availability":"SPOT_WITH_FALLBACK",                                \
        "spot_bid_price_percent":100                                        \
     },                                                                     \
   "custom_tags":{                                                          \
        "ResourceClass":"Serverless"                                        \
     },                                                                     \
       "autoscale":{                                                        \
        "min_workers":1,                                                    \
        "max_workers":2                                                     \
     },                                                                     \
  "autotermination_minutes":10                                              \
}' https://dogfood.staging.cloud.databricks.com/api/2.0/clusters/create '

### Answer:

`Network Isolation - create a dedicated VPC/VNET in the customer’s account that only contains Databricks infrastructure. This ensures that all Databricks policies and controls have no access to other production infrastructure in a customers account. Additionally, all communication with Databricks’ control infrastructure goes through direct links (while additionally eliminating any Public IPs). This prevents any traffic from traversing the public internet and is also encrypted with mutual TLS v1.2. And finally, multiple security groups are used so that if ever Databricks services need to connect to other customer VPC/VNETs (e.g., containing another production data source), there is no risk of exposing access to those services to the outside world.

Enable SSO - Integration with SSO identity providers (such as Okta, Onelogin etc.) and support for SCIM allow you to easily manage and secure users.

Enable ACLs - Cluster, Data, job, Workspace level

Enable encryption at rest(storage level) and in motion(using custom code).`

# This is the end of the official test. Bonus below!

## Part 7: Bonus: Data Science & Machine Learning

### Question 16: Conceptual Question - A Skewed Feature

One of these lines is the *mean* of this feature. The other is the *median*. Which of these lines is the **mean** - the red line or the black line?

<img width=400px src=https://www.evernote.com/l/AAEycL6CQ0hLi5V5pIo91Ko-Pfk2i0AnGyMB/image.png>

### Answer:

`The mean is always greater than the median, hence, red line is the median and black is the median.`

### Question 17: Conceptual Question - Exploratory Data Analysis

The plots below show the distribution of home selling prices differentiated by a few categorical features. Based on these plots, **which of these categorical features** - Property Type, Exterior Quality, or Month Sold - would you expect to be most associated with Price? **Why**?

<img width=1600px src=https://www.evernote.com/l/AAHulkcc20hHSJV6D1udKiwSDCN0S6oV_5YB/image.png>

### Answer:

The Exterior quality feature seems to be most associated with price as we can see the given graph

### Question 18: Conceptual Question - Analyze Model Performance

Consider the following results for a decision tree model against training and testing data sets:

`decision tree regression - train r2 score: 0.9944`  
`decision tree regression - test r2 score:  0.3119`


What is your assessment of this model?

### Answer:
The model is highly over-fitted because the training accuracy is very high as compared to the testing accuracy, we need to retrain the model and do cross validation techniques to verify our accuracy

### Question 19: Conceptual Question - Model Selection

A series of models has been built using the same training data, but each with a subset of features.

Consider the following results for a series of logistic regression models and then answer this question:

- Which model would you choose and why?
- What other things would you want to look at and why?

| model number | feature subset | logistic regression test accuracy|
|:-:|:-:|:-:|
| 1| feat_1 |	 0.631|
| 2| feat_2 |	 0.552|
| 3| feat_3 |	 0.868|
| 4| feat_4 |	 0.868|
| 5| feat_1, feat_2 |	 0.657|
| 6| feat_1, feat_3 |	 0.947|
| 7| feat_1, feat_4 |	 0.921|
| 8| feat_2, feat_3 |	 0.947|
| 9| feat_2, feat_4 |	 0.973|
| 10| feat_3, feat_4 |	 0.947|
| 11| feat_1, feat_2, feat_3 |	 0.947|
| 12| feat_1, feat_2, feat_4 |	 0.947|
| 13| feat_1, feat_3, feat_4 |	 0.947|
| 14| feat_2, feat_3, feat_4 |	 0.973|
| 15| feat_1, feat_2, feat_3, feat_4 |	 0.973|

### Answer:
I would chose model number 14 because it is giving highest accuracy and also it is better than model number 15 because it has an extra feature less which is not affecting the accuracy

-sandbox
&copy; 2019 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>